packages/mongodb-rag-core/src/contentStore/updateEmbeddedContent.ts

```diff
@@ -48,6 +63,18 @@ export const updateEmbeddedContent = async ({
       sourceNames ? ` in sources: ${sourceNames.join(", ")}` : ""
     }`
   );
+  // find pages with changed chunkingHashes
+  const allPages = await pageStore.loadPages({sources: sourceNames});
+  const pagesWithChangedChunking = allPages.filter(async (page) => {
+    const existingContent = await embeddedContentStore.loadEmbeddedContent({
+      page,
+    });
+    const chunkAlgoHash = getHashForFunc(chunkPage, chunkOptions);
+    return existingContent[0].chunkAlgoHash !== chunkAlgoHash
+  });
+
+  changedPages.push(...pagesWithChangedChunking);
```

Comment:
this is very inefficient and uses a lot of memory b/c it loads all pages, loops thru them one by one. pls reconsider a different implementation

Reply:
i think you can use the DB more here to make this logic happen at the DB level. 

you could: 


add a method to the embedded content collection that finds the data sources for a given query, eg:

```ts

// typing 
type EmbeddedContentStore {
  getDataSources(matchQuery: Document): Promise<string[]>
}

// implementation w mongodb
{
  async getDataSources(matchQuery) {
    const {uniqueSources } = await embeddedContentStore.aggregate([

  /* example match query:
{
    "$match": {
      "chunkAlgoHash": "49d78a1d6b12ee6f433a2156060ed5ebfdefb8a90301f1c2fb04e4524944c5eb",
      "sourceName": {
        "$in": [
          "snooty-cloud-docs",
          "snooty-docs"
        ]
      }
    }
  },

  */
  { $match: matchQuery },
  {
    "$group": {
      "_id": null,
      "uniqueSources": {
        "$addToSet": "$sourceName"
      }
    }
  },
  {
    "$project": {
      "_id": 0,
      "uniqueSources": 1
    }
  }
]).toArray();
return uniqueSources;
  }
}

```

you can then merge this list of sources that require update before the initial call to pageStore.loadPages() on line 57:


pseudocode:

```ts
const chunkAlgoHashes = new Map<string, string>();
const chunkAlgoHash = getHashForFunc(chunkPage, chunkOptions, chunkAlgoHashes);

const sourcesWithNewChunkAlgoHash = await embeddedContentStore.getDataSources({
      chunkAlgoHash,
      // just run on specific source names if specified, run on all if not
      ...(sourceNames ? {"sourceName": {
        "$in": sourceNames
      }}: undefined)
    })

const allSourcesToUpdate = new Array(new Set([...sourcesWithNewChunkAlgoHash, sourceNames]])); //smthn like this..not 100% sure on syntax
const changedPages = await pageStore.loadPages({
    updated: since,
    sources: allSourcesToUpdate,
  });

// ... rest of code unchanged

```

Reply:
you may also want to add an index to 'chunkAlgoHash' on the embeddedContent collection to improve query performance (not that it particularly matters here since this is just run once per CLI inovcation)

Reply:
note that this implementation wouldn't be sufficient if we allowed for updating chunks based on criteria besides `dataSource`...but that's not how the CLI currently works, so this implementation should get the job done

Reply:
refactored to get pages based on embedded content with an aggregation pipeline that uses a $lookup